### PR TITLE
[Autograph] Fix loop else support and add loop integration test

### PR DESCRIPTION
**Pull request by zhuzilin**
This is a PR from JIZHI, the AI platform in Tencent.

Right now, the autograph does not support loop else syntax like
```python
while x > 2:
  x /= 2
else:
  x += 1
```
In fact, it would just neglect the else part, because the tf functional while op does not support "else" (which, in my opinion, is good, since the loop else is a rare syntax sugar that probably only appears in python). Therefore, this pr changed the loop from
```python
while test:
  body
else:
  orelse
```
to
```python
while test:
  body
orelse
```
And for those loops that have `break` statements involve
```python
var_name = False
while ag__.and_(lambda: test, lambda: ag__.not_(var_name)):
  body
if ag__.not_(var_name):
  orelse
```
For the test, it turns out that the UT for break statement only could not detect this problem. In fact, the `test_loop_orelse` in `break_statement_test.py` does not work in current autograph but have passed the UT. Therefore, we added a integration test for the loop syntax, which will run the break_statement, continue_statement and control_flow converters one by one.

Thank you for your time on this review.

### CODE REVIEW

File: tensorflow/python/autograph/converters/break_statements.py [hunk 1]
```diff
@@ -101,6 +100,17 @@ def visit_While(self, node):
 
       new_while_node = node[1]
       anno.copyanno(original_node, new_while_node, anno.Basic.DIRECTIVES)
```
Comment (mdanatg) [inline tensorflow/python/autograph/converters/break_statements.py]: This annotation will need to be copied on the else branch as well (same for the for loop case).

File: tensorflow/python/autograph/converters/break_statements.py [hunk 2]
```diff
@@ -101,6 +100,17 @@ def visit_While(self, node):
...
+    else:
```
Comment (mdanatg) [inline tensorflow/python/autograph/converters/break_statements.py]: The code will be a bit easier to follow if written like so:

```
if not break_used:
  <<apply the simple template>>
  return node

<<apply the main template>>
```

File: tensorflow/python/autograph/converters/break_statements.py [hunk 3]
```diff
@@ -145,7 +154,18 @@ def visit_For(self, node):
       new_for_node = node[1]
       anno.setanno(new_for_node, 'extra_test', extra_test)
```
Comment (mdanatg) [inline tensorflow/python/autograph/converters/break_statements.py]: These two annotations need to be copied over (use `anno.copyanno` in both cases)

File: tensorflow/python/autograph/converters/loop_integration_test.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        if x % 2 == 0:
+          break
+        v.append(x)
+      return v
+
+    self.assertTransformedEquivalent(test_fn, 0)
+    self.assertTransformedEquivalent(test_fn, 1)
+    self.assertTransformedEquivalent(test_fn, 4)
+
+  def test_for_loop(self):
```
Comment (mdanatg) [inline tensorflow/python/autograph/converters/loop_integration_test.py]: These tests seem identical to `break_statements_test.py` - what are the differences? Can you add the extra test cases in there instead?

File: tensorflow/python/autograph/converters/loop_integration_test.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        x -= 1
+        if x % 2 == 0:
+          break
+        v.append(x)
+      return v
+
+    with self.converted(test_fn, break_statements, {},
+                        (constant_op.constant,)) as result:
+      # The break is incompletely canonicalized. The loop will not interrupt,
+      # but the section following the break will be skipped.
```
Comment (mdanatg) [inline tensorflow/python/autograph/converters/loop_integration_test.py]: I think the PR is meant to fix exactly this issue, is it not?

**Review by mdanatg — CHANGES_REQUESTED** [review]
Thank you for the fix. I hadn't realized that this is the only missing link, since `control_flow.py` will lower the guarded orelse into the appropriate if statement. Nice work!

File: tensorflow/python/autograph/converters/break_statements.py (hunk 1 above)
Comment (zhuzilin) [reply tensorflow/python/autograph/converters/break_statements.py]: Added.

File: tensorflow/python/autograph/converters/break_statements.py (hunk 2 above)
Comment (zhuzilin) [reply tensorflow/python/autograph/converters/break_statements.py]: I have adjusted the structure.

File: tensorflow/python/autograph/converters/break_statements.py (hunk 3 above)
Comment (zhuzilin) [reply tensorflow/python/autograph/converters/break_statements.py]: Added.

File: tensorflow/python/autograph/converters/loop_integration_test.py (hunk 4 above)
Comment (zhuzilin) [reply tensorflow/python/autograph/converters/loop_integration_test.py]: The reason for adding those two tests is to make sure the normal loops can successfully go through all the converters. You are right, they are probably redundant and I have removed them.

File: tensorflow/python/autograph/converters/loop_integration_test.py (hunk 5 above)
Comment (zhuzilin) [reply tensorflow/python/autograph/converters/loop_integration_test.py]: I'm really sorry about this... I just copied the test from break_statement_test.py for the reason I mentioned above and didn't notice the extra self.converted... Anyway, those two unnecessary tests are removed...

**Review by zhuzilin — COMMENTED** [review]
@mdanatg Thank you for your quick review! I've updated the PR according to the comments. Could you have a second look?

File: tensorflow/python/autograph/converters/break_statements.py [hunk 6]
```diff
@@ -115,37 +130,53 @@ def visit_For(self, node):
...
           iter_=node.iter,
           target=node.target,
           body=node.body,
-          orelse=guarded_orelse)
+          orelse=node.orelse)
 
-      new_for_node = node[1]
-      anno.setanno(new_for_node, anno.Basic.EXTRA_LOOP_TEST, extra_test)
+      new_for_node = node[0]
       anno.copyanno(original_node, new_for_node, anno.Basic.DIRECTIVES)
```
Comment (mdanatg) [inline tensorflow/python/autograph/converters/break_statements.py]: Just to be safe, please also copy the "extra_test" annotation. Thanks!

File: tensorflow/python/autograph/converters/loop_integration_test.py (hunk 5 above)
Comment (mdanatg) [reply tensorflow/python/autograph/converters/loop_integration_test.py]: Ah, I see. It strikes me that break_statement_test.py now has a comment that is stale - feel free to update it as well.

**Review by mdanatg — DISMISSED** [review]

File: tensorflow/python/autograph/converters/break_statements.py (hunk 6 above)
Comment (zhuzilin) [reply tensorflow/python/autograph/converters/break_statements.py]: Added.

**Review by mdanatg — APPROVED** [review]

**Comment by zhuzilin** [comment]
@mdanatg Thank you so much for your quick response! I haven't even finish my comment... As for the comment in break statement, break in for loop will still take effect only after the control_flow converter, so I think the UT makes sense and therefore didn't touch it.

**Comment by zhuzilin** [comment]
@mdanatg Could you help me merge this pr? thx!

**Comment by mdanatg** [comment]
Sorry for the delay. The PR tripped a few internal tests. It looks like some changes got lost in the merge.

File: tensorflow/python/autograph/converters/break_statements.py [hunk 7]
```diff
@@ -115,37 +130,54 @@ def visit_For(self, node):
...
           iter_=node.iter,
           target=node.target,
           body=node.body,
-          orelse=guarded_orelse)
+          orelse=node.orelse)
 
-      new_for_node = node[1]
-      anno.setanno(new_for_node, anno.Basic.EXTRA_LOOP_TEST, extra_test)
+      new_for_node = node[0]
+      anno.copyanno(original_node, new_for_node, 'extra_test')
```
Comment (mdanatg) [inline tensorflow/python/autograph/converters/break_statements.py]: Please keep the `anno.Basic.EXTRA_LOOP_TEST` here.

File: tensorflow/python/autograph/converters/break_statements.py [hunk 8]
```diff
@@ -115,37 +130,54 @@ def visit_For(self, node):
...
+    node = templates.replace(
+        template,
+        var_name=break_var,
+        iter_=node.iter,
+        target=node.target,
+        body=node.body,
+        orelse=guarded_orelse)
+
+    new_for_node = node[1]
+    anno.setanno(new_for_node, 'extra_test', extra_test)
```
Comment (mdanatg) [inline tensorflow/python/autograph/converters/break_statements.py]: Use `anno.Basic.EXTRA_LOOP_TEST` here.

**Review by mdanatg — CHANGES_REQUESTED** [review]

File: tensorflow/python/autograph/converters/break_statements.py (hunk 7 above)
Comment (zhuzilin) [reply tensorflow/python/autograph/converters/break_statements.py]: changed.

File: tensorflow/python/autograph/converters/break_statements.py (hunk 8 above)
Comment (zhuzilin) [reply tensorflow/python/autograph/converters/break_statements.py]: changed.

**Review by mdanatg — APPROVED** [review]
Thanks!

**Pull request merged**